### PR TITLE
plugin track & playlist json modifications

### PR DESCRIPTION
**Pull request by topi314**
This a draft pr & open for discussion

this pr adds the ability for plugins to add custom fields to the audio track info & playlist info json
This is built on top of the new rest api and therefore `feature/rest-api` is it's base

### CODE REVIEW

**Comment by topi314** [comment]
this is usable & testable but misses documentation
I already implemented this for all sources in my `LavaSrc` plugin which can be therefore used for testing. See https://github.com/TopiSenpai/LavaSrc/pull/36

**Comment by topi314** [comment]
Things to consider in terms of code:
 * Make one JsonAppender interface instead of AudioTrack & AudioPlaylist interfaces

**Comment by topi314** [comment]
I got some feedback that something like this
```json
{
  "encoded": "...",
  "info": {
    "identifier": "...."
  },
  "plugin": {
    "artworkURL": "..."
  }
}
```
Would be easier to parse for libraries than this (current):
```json
{
  "encoded": "...",
  "info": {
    "identifier": "....",
    "artworkURL": "..."
  }
}
```

Any thoughts?

**Comment by topi314** [comment]
for playlists we could deprecate `playlistInfo` and do:
```json
{
  "playlist": {
    "info": {
      "name": "...",
      "selectedTrack": 0
    },
    "plugin": {
      "artworkURL": "..."
    }
  }
}
```
current:
```json
{
  "playlistInfo": {
    "name": "...",
    "selectedTrack": 0,
    "artworkURL": "..."
  }
}
```

**Comment by duncte123** [comment]
> I got some feedback that something like this
> 
> ```json
> {
>   "encoded": "...",
>   "info": {
>     "identifier": "...."
>   },
>   "plugin": {
>     "artworkURL": "..."
>   }
> }
> ```
> 
> Would be easier to parse for libraries than this (current):
> 
> ```json
> {
>   "encoded": "...",
>   "info": {
>     "identifier": "....",
>     "artworkURL": "..."
>   }
> }
> ```
> 
> Any thoughts?

Massive preference for the first one from me. Do the same with playlists since consistency is good

**Comment by freyacodes** [comment]
Don't merge this until we have a release of v3.8

I think we should consider whether we should instead let plugins provide data through a new JSON object (e.g. called `pluginData`) instead of letting them edit the main track info

**Comment by topi314** [comment]
> Don't merge this until we have a release of v3.8
> 
> I think we should consider whether we should instead let plugins provide data through a new JSON object (e.g. called `pluginData`) instead of letting them edit the main track info

editing the track json was my initial thought & that's how it was originally implemented in https://github.com/freyacodes/Lavalink/pull/618

however I don't see why a plugin would need to edit already existing fields. If we look at how the encoded track works in lavaplayer the source managers are only able to add custom fields but not edit existing. As this is essentially the same feature I think `pluginData` is sufficient.

**Comment by topi314** [comment]
to make naming more consistent I think `pluginInfo` makes the most sense

**Comment by topi314** [comment]
the new json looks as following:
```json
{
	"encoded": "...",	
	"info": {
		"identifier": "2ZmjG3wuUes",
		"isSeekable": true,
		"author": "Epitaph Records",
		"length": 243000,
		"isStream": false,
		"position": 0,
		"title": "Architects - \"Dead Butterflies\"",
		"uri": "https://www.youtube.com/watch?v=2ZmjG3wuUes",
		"sourceName": "youtube"
	,
	"pluginInfo": {
		"artworkUrl": "https://img.youtube.com/vi/2ZmjG3wuUes/maxresdefault.jpg"
	}
}	
```

```json
{
	"loadType": "PLAYLIST_LOADED",
	"playlistInfo": {
		"name": "Architects - the classic symptoms of a broken spirit",
		"selectedTrack": -1
	},
	"pluginInfo": {
		"artworkUrl": "https://img.youtube.com/vi/2ZmjG3wuUes/maxresdefault.jpg"
	},
	"tracks": []
}
```

**Comment by Drapersniper** [comment]
Food for thought - given that localtracks add the probe info, the http tracks add one additional key.

would it make sense to move these "extras" to the pluginInfo field ? Maybe make it "extraInfo" or "sourceExtras"?  That way there could be a canonical place for any info that doesn't belong inside "info" as standard ? (By this I mean any info that isn't available across all sources by default)

**Comment by topi314** [comment]
that info would also belong to `pluginInfo` imo and would also be added by a plugin. It may or may not be bundled with lavalink

**Comment by Drapersniper** [comment]
Attaching some of the loadtrack responses with the proposed changes for reference
`/v4/loadtracks`
```json
{
    "loadType": "TRACK_LOADED",
    "tracks": [
        {
            "encoded": "QAABCAMAKUNsZWFuIFdoaXRlIE5vaXNlIC0gTG9vcGFibGUgd2l0aCBubyBmYWRlABZXaGl0ZSBOb2lzZSBCYWJ5IFNsZWVwAAAAAAABX5AACTU3MTYwNzc2MgABACJodHRwczovL2RlZXplci5jb20vdHJhY2svNTcxNjA3NzYyAQBpaHR0cHM6Ly9lLWNkbnMtaW1hZ2VzLmR6Y2RuLm5ldC9pbWFnZXMvY292ZXIvZDFlMzJjY2RiY2YxZmUzNTY5ZTEyYzNjNGMyZDY4NjEvMTAwMHgxMDAwLTAwMDAwMC04MC0wLTAuanBnAQAMU0VZT0wxNzk3MjAxAAZkZWV6ZXIAAAAAAAAAAA==",
            "info": {
                "identifier": "571607762",
                "isSeekable": true,
                "author": "White Noise Baby Sleep",
                "length": 90000,
                "isStream": false,
                "position": 0,
                "title": "Clean White Noise - Loopable with no fade",
                "uri": "https://deezer.com/track/571607762",
                "sourceName": "deezer",
                "artworkUrl": "https://e-cdns-images.dzcdn.net/images/cover/d1e32ccdbcf1fe3569e12c3c4c2d6861/1000x1000-000000-80-0-0.jpg",
                "isrc": "SEYOL1797201"
            },
            "pluginInfo": {}
        }
    ],
    "playlistInfo": null,
    "pluginInfo": null,
    "exception": null
}
```


```json
{
    "loadType": "PLAYLIST_LOADED",
    "tracks": [
        {
            "encoded": "QAAA5AMAH0FsbCBJIFdhbnQgZm9yIENocmlzdG1hcyBJcyBZb3UADE1hcmlhaCBDYXJleQAAAAAAA61oAAg2MjcxMDQ0MgABACFodHRwczovL2RlZXplci5jb20vdHJhY2svNjI3MTA0NDIBAGlodHRwczovL2UtY2Rucy1pbWFnZXMuZHpjZG4ubmV0L2ltYWdlcy9jb3Zlci82MGQ4Zjc2MzVlYWQ5YjNjNWU1NzFhZWM0N2E4YzczZS8xMDAweDEwMDAtMDAwMDAwLTgwLTAtMC5qcGcAAAZkZWV6ZXIAAAAAAAAAAA==",
            "info": {
                "identifier": "62710442",
                "isSeekable": true,
                "author": "Mariah Carey",
                "length": 241000,
                "isStream": false,
                "position": 0,
                "title": "All I Want for Christmas Is You",
                "uri": "https://deezer.com/track/62710442",
                "sourceName": "deezer",
                "artworkUrl": "https://e-cdns-images.dzcdn.net/images/cover/60d8f7635ead9b3c5e571aec47a8c73e/1000x1000-000000-80-0-0.jpg",
                "isrc": null
            },
            "pluginInfo": {}
        },
        ...
    ],
    "playlistInfo": {
        "name": "Top UK",
        "selectedTrack": -1
    },
    "pluginInfo": {
        "type": "playlist",
        "identifier": "1111142221",
        "artworkURL": "https://e-cdns-images.dzcdn.net/images/playlist/85bd6585976b8e21e0deee810ff24518/1000x1000-000000-80-0-0.jpg",
        "author": "Deezer Charts"
    },
    "exception": null
}
```
```json
{
    "loadType": "PLAYLIST_LOADED",
    "tracks": [
        {
            "encoded": "QAABFAMAe0dvZCBrbm93cy4uLiAnJ1RoZSBNZWxhbmNob2x5IG9mIEhhcnVoaSBTdXp1bWl5YScnIOOAkOa2vOWuruODj+ODq+ODkuOBruaGgumsseOAkUthZG9rYXdh5YWs6KqNTUFE44CQ776N776e772w7729IOa8lOWlj+OAkQALU09TIEJyaWdhZGUAAAAAAARJqAALV1dCMDFJdU12ekEAAQAraHR0cHM6Ly93d3cueW91dHViZS5jb20vd2F0Y2g/dj1XV0IwMUl1TXZ6QQEAMGh0dHBzOi8vaS55dGltZy5jb20vdmkvV1dCMDFJdU12ekEvbXFkZWZhdWx0LmpwZwAAB3lvdXR1YmUAAAAAAAAAAA==",
            "info": {
                "identifier": "WWB01IuMvzA",
                "isSeekable": true,
                "author": "SOS Brigade",
                "length": 281000,
                "isStream": false,
                "position": 0,
                "title": "God knows... ''The Melancholy of Haruhi Suzumiya'' 【涼宮ハルヒの憂鬱】Kadokawa公認MAD【ﾍﾞｰｽ 演奏】",
                "uri": "https://www.youtube.com/watch?v=WWB01IuMvzA",
                "sourceName": "youtube",
                "artworkUrl": "https://i.ytimg.com/vi/WWB01IuMvzA/mqdefault.jpg",
                "isrc": null
            },
            "pluginInfo": {}
        },
        ...
    ],
    "playlistInfo": {
        "name": "Best anime songs playlist",
        "selectedTrack": -1
    },
    "pluginInfo": {},
    "exception": null
}
```
```json
{
    "loadType": "SEARCH_RESULT",
    "tracks": [
        {
            "encoded": "QAAApAMADUFkZWxlIC0gSGVsbG8ABUFkZWxlAAAAAAAFmZgAC1lRSHNYTWdsQzlBAAEAK2h0dHBzOi8vd3d3LnlvdXR1YmUuY29tL3dhdGNoP3Y9WVFIc1hNZ2xDOUEBADRodHRwczovL2kueXRpbWcuY29tL3ZpL1lRSHNYTWdsQzlBL21heHJlc2RlZmF1bHQuanBnAAAHeW91dHViZQAAAAAAAAAA",
            "info": {
                "identifier": "YQHsXMglC9A",
                "isSeekable": true,
                "author": "Adele",
                "length": 367000,
                "isStream": false,
                "position": 0,
                "title": "Adele - Hello",
                "uri": "https://www.youtube.com/watch?v=YQHsXMglC9A",
                "sourceName": "youtube",
                "artworkUrl": "https://i.ytimg.com/vi/YQHsXMglC9A/maxresdefault.jpg",
                "isrc": null
            },
            "pluginInfo": {}
        },
        ...
    ],
    "playlistInfo": null,
    "pluginInfo": null,
    "exception": null
}
```
```json
{
    "loadType": "SEARCH_RESULT",
    "tracks": [
        {
            "encoded": "QAAAxQMABUhlbGxvAAVFcmF0bwAAAAAAA4JwAAkxMTAzMTU5NjIAAQAiaHR0cHM6Ly9kZWV6ZXIuY29tL3RyYWNrLzExMDMxNTk2MgEAaWh0dHBzOi8vZS1jZG5zLWltYWdlcy5kemNkbi5uZXQvaW1hZ2VzL2NvdmVyLzMxYzYxM2NlZWE2OTQ5NTliOTEyMTAzNDY3YmRhM2FiLzEwMDB4MTAwMC0wMDAwMDAtODAtMC0wLmpwZwAABmRlZXplcgAAAAAAAAAA",
            "info": {
                "identifier": "110315962",
                "isSeekable": true,
                "author": "Erato",
                "length": 230000,
                "isStream": false,
                "position": 0,
                "title": "Hello",
                "uri": "https://deezer.com/track/110315962",
                "sourceName": "deezer",
                "artworkUrl": "https://e-cdns-images.dzcdn.net/images/cover/31c613ceea694959b912103467bda3ab/1000x1000-000000-80-0-0.jpg",
                "isrc": null
            },
            "pluginInfo": {}
        },
        ...
    ],
    "playlistInfo": null,
    "pluginInfo": null,
    "exception": null
}
```

```json
{
    "loadType": "TRACK_LOADED",
    "tracks": [
        {
            "encoded": "QAAA3AMAFEFuZ2VsIFdpdGggYSBTaG90Z3VuAAdUaGUgQ2FiAAAAAAADZ3QAFjVBOGpVWmF0aXg4N3hHMFpqNkhTS0cAAQA1aHR0cHM6Ly9vcGVuLnNwb3RpZnkuY29tL3RyYWNrLzVBOGpVWmF0aXg4N3hHMFpqNkhTS0cBAEBodHRwczovL2kuc2Nkbi5jby9pbWFnZS9hYjY3NjE2ZDAwMDBiMjczZjliYjIwNGNjNzUxZWYxOTYwYjFlZWM2AQAMVVNKNVQxMTAwMTcwAAdzcG90aWZ5AAAAAAAAAAA=",
            "info": {
                "identifier": "5A8jUZatix87xG0Zj6HSKG",
                "isSeekable": true,
                "author": "The Cab",
                "length": 223092,
                "isStream": false,
                "position": 0,
                "title": "Angel With a Shotgun",
                "uri": "https://open.spotify.com/track/5A8jUZatix87xG0Zj6HSKG",
                "sourceName": "spotify",
                "artworkUrl": "https://i.scdn.co/image/ab67616d0000b273f9bb204cc751ef1960b1eec6",
                "isrc": "USJ5T1100170"
            },
            "pluginInfo": {}
        }
    ],
    "playlistInfo": null,
    "pluginInfo": null,
    "exception": null
}
```
```json
{
    "loadType": "PLAYLIST_LOADED",
    "tracks": [
        {
            "encoded": "QAAA3AMAFEFuZ2VsIFdpdGggYSBTaG90Z3VuAAdUaGUgQ2FiAAAAAAADZ3QAFjVBOGpVWmF0aXg4N3hHMFpqNkhTS0cAAQA1aHR0cHM6Ly9vcGVuLnNwb3RpZnkuY29tL3RyYWNrLzVBOGpVWmF0aXg4N3hHMFpqNkhTS0cBAEBodHRwczovL2kuc2Nkbi5jby9pbWFnZS9hYjY3NjE2ZDAwMDBiMjczZjliYjIwNGNjNzUxZWYxOTYwYjFlZWM2AQAMVVNKNVQxMTAwMTcwAAdzcG90aWZ5AAAAAAAAAAA=",
            "info": {
                "identifier": "5A8jUZatix87xG0Zj6HSKG",
                "isSeekable": true,
                "author": "The Cab",
                "length": 223092,
                "isStream": false,
                "position": 0,
                "title": "Angel With a Shotgun",
                "uri": "https://open.spotify.com/track/5A8jUZatix87xG0Zj6HSKG",
                "sourceName": "spotify",
                "artworkUrl": "https://i.scdn.co/image/ab67616d0000b273f9bb204cc751ef1960b1eec6",
                "isrc": "USJ5T1100170"
            },
            "pluginInfo": {}
        },
        ...
    ],
    "playlistInfo": {
        "name": "???",
        "selectedTrack": -1
    },
    "pluginInfo": {
        "type": "playlist",
        "identifier": "0BbMjMQZ43vtdz7al266XH",
        "artworkURL": "https://mosaic.scdn.co/640/ab67616d0000b2731bfa1a433edde0dedb26e0b7ab67616d0000b27374164dc6fb6a320abf6a34fdab67616d0000b273b11c3ab5cb818192b7a61fddab67616d0000b273f9bb204cc751ef1960b1eec6",
        "author": null
    },
    "exception": null
}
```
```json
{
    "loadType": "NO_MATCHES",
    "tracks": [],
    "playlistInfo": null,
    "pluginInfo": null,
    "exception": null
}
```
`/v4/decodetrack`
```json
{
    "encoded": "QAAA3AMAFEFuZ2VsIFdpdGggYSBTaG90Z3VuAAdUaGUgQ2FiAAAAAAADZ3QAFjVBOGpVWmF0aXg4N3hHMFpqNkhTS0cAAQA1aHR0cHM6Ly9vcGVuLnNwb3RpZnkuY29tL3RyYWNrLzVBOGpVWmF0aXg4N3hHMFpqNkhTS0cBAEBodHRwczovL2kuc2Nkbi5jby9pbWFnZS9hYjY3NjE2ZDAwMDBiMjczZjliYjIwNGNjNzUxZWYxOTYwYjFlZWM2AQAMVVNKNVQxMTAwMTcwAAdzcG90aWZ5AAAAAAAAAAA=",
    "info": {
        "identifier": "5A8jUZatix87xG0Zj6HSKG",
        "isSeekable": true,
        "author": "The Cab",
        "length": 223092,
        "isStream": false,
        "position": 0,
        "title": "Angel With a Shotgun",
        "uri": "https://open.spotify.com/track/5A8jUZatix87xG0Zj6HSKG",
        "sourceName": "spotify",
        "artworkUrl": "https://i.scdn.co/image/ab67616d0000b273f9bb204cc751ef1960b1eec6",
        "isrc": "USJ5T1100170"
    },
    "pluginInfo": {}
}
```
`/v4/decodetracks`
```json
[
    {
        "encoded": "QAAA3AMAFEFuZ2VsIFdpdGggYSBTaG90Z3VuAAdUaGUgQ2FiAAAAAAADZ3QAFjVBOGpVWmF0aXg4N3hHMFpqNkhTS0cAAQA1aHR0cHM6Ly9vcGVuLnNwb3RpZnkuY29tL3RyYWNrLzVBOGpVWmF0aXg4N3hHMFpqNkhTS0cBAEBodHRwczovL2kuc2Nkbi5jby9pbWFnZS9hYjY3NjE2ZDAwMDBiMjczZjliYjIwNGNjNzUxZWYxOTYwYjFlZWM2AQAMVVNKNVQxMTAwMTcwAAdzcG90aWZ5AAAAAAAAAAA=",
        "info": {
            "identifier": "5A8jUZatix87xG0Zj6HSKG",
            "isSeekable": true,
            "author": "The Cab",
            "length": 223092,
            "isStream": false,
            "position": 0,
            "title": "Angel With a Shotgun",
            "uri": "https://open.spotify.com/track/5A8jUZatix87xG0Zj6HSKG",
            "sourceName": "spotify",
            "artworkUrl": "https://i.scdn.co/image/ab67616d0000b273f9bb204cc751ef1960b1eec6",
            "isrc": "USJ5T1100170"
        },
        "pluginInfo": {}
    },
    {
        "encoded": "QAAA5AMAH0FsbCBJIFdhbnQgZm9yIENocmlzdG1hcyBJcyBZb3UADE1hcmlhaCBDYXJleQAAAAAAA61oAAg2MjcxMDQ0MgABACFodHRwczovL2RlZXplci5jb20vdHJhY2svNjI3MTA0NDIBAGlodHRwczovL2UtY2Rucy1pbWFnZXMuZHpjZG4ubmV0L2ltYWdlcy9jb3Zlci82MGQ4Zjc2MzVlYWQ5YjNjNWU1NzFhZWM0N2E4YzczZS8xMDAweDEwMDAtMDAwMDAwLTgwLTAtMC5qcGcAAAZkZWV6ZXIAAAAAAAAAAA==",
        "info": {
            "identifier": "62710442",
            "isSeekable": true,
            "author": "Mariah Carey",
            "length": 241000,
            "isStream": false,
            "position": 0,
            "title": "All I Want for Christmas Is You",
            "uri": "https://deezer.com/track/62710442",
            "sourceName": "deezer",
            "artworkUrl": "https://e-cdns-images.dzcdn.net/images/cover/60d8f7635ead9b3c5e571aec47a8c73e/1000x1000-000000-80-0-0.jpg",
            "isrc": null
        },
        "pluginInfo": {}
    }
]
```
`/v4/sessions/{SESSION_ID}/players/{GUILD_ID}`
```json
{
    "guildId": "133049272517001216",
    "track": {
        "encoded": "QAAAyAMAAklEAAlDaWRlcmdpcmwAAAAAAANu6AAKMTAwNTY4NDkzMgABACNodHRwczovL2RlZXplci5jb20vdHJhY2svMTAwNTY4NDkzMgEAaWh0dHBzOi8vZS1jZG5zLWltYWdlcy5kemNkbi5uZXQvaW1hZ2VzL2NvdmVyLzcxNGUyNmM1Mjg1MzdlNTc2Yzg1ZTMzYWYzMjI2OWY2LzEwMDB4MTAwMC0wMDAwMDAtODAtMC0wLmpwZwAABmRlZXplcgAAAAAAA04E",
        "info": {
            "identifier": "1005684932",
            "isSeekable": true,
            "author": "Cidergirl",
            "length": 225000,
            "isStream": false,
            "position": 216580,
            "title": "ID",
            "uri": "https://deezer.com/track/1005684932",
            "sourceName": "deezer",
            "artworkUrl": "https://e-cdns-images.dzcdn.net/images/cover/714e26c528537e576c85e33af32269f6/1000x1000-000000-80-0-0.jpg",
            "isrc": null
        },
        "pluginInfo": {}
    },
    "volume": 100,
    "paused": false,
    "voice": {},
    "filters": {}
}```

File: IMPLEMENTATION.md [hunk 1]
```diff
@@ -605,10 +605,11 @@ GET /v4/sessions/{sessionId}/players
...
-|---------|----------------------------------|--------------------------------------------------------------------------------------|
-| encoded | string                           | The base64 encoded track data                                                        |
-| info    | [Track Info](#track-info) object | Info about the track                                                                 |
+| Field      | Type                             | Description                                                                          |
+|------------|----------------------------------|--------------------------------------------------------------------------------------|
+| encoded    | string                           | The base64 encoded track data                                                        |
+| info       | [Track Info](#track-info) object | Info about the track                                                                 |
+| pluginInfo | object                           | Additional track info from plugins                                                   |
```
Comment (Drapersniper) [inline IMPLEMENTATION.md]: Documentation for WS changes are currently incomplete (Missing pluginInfo)

**Comment by Tomato6966** [comment]
Already implemented it in my warpper
-> Suggestion: don't use maxresdefault ---> cause youtube does not always provide maxresdefault .

The only thumbnail that's always available + never has black bars is: mqdefault

I just replace the maxresdefault with mqdefault, however if you'd do that that would ensure that it's more reliable.

**Comment by topi314** [comment]
That is not in my hands. We use lavaplayer and lavaplayer provides the thumbnails

**Comment by Tomato6966** [comment]
> That is not in my hands. We use lavaplayer and lavaplayer provides the thumbnails

I know, you could just replace maxresdefault.jpg with mqdefault.jpg in the artworkURL string
-> if you as Lavalink don't want this --> than the ones who make clients / use it may read this and now know how to fix this problem that the image does not show, tho the link is given.

**Comment by Drapersniper** [comment]
> > That is not in my hands. We use lavaplayer and lavaplayer provides the thumbnails
> 
> I know, you could just replace maxresdefault.jpg with mqdefault.jpg in the artworkURL string -> if you as Lavalink don't want this --> than the ones who make clients / use it may read this and now know how to fix this problem that the image does not show, tho the link is given.

You should be making the suggestion agaisnt this function

https://github.com/Walkyst/lavaplayer-fork/blob/custom/main/src/main/java/com/sedmelluq/discord/lavaplayer/tools/PBJUtils.java

**Pull request merged**